### ISSUE TITLE
Playing Catch

### ISSUE THREAD

**Issue by xEvoGx** [thestonefox/SteamVR_Unity_Toolkit#94]
Hey Stone! I mentioned this way back in the thread regarding the door and joint contraints...

...but catching objects you toss up is difficult the higher it is. Naturally IRL, you're able to get your hand _in front_ of a falling object so you can (ideally) catch it. Right now with your current implementation, grabbing only occurs when the controller is touching the object. Trying to catch a falling object where you have to pull the trigger **right** at the moment they're touching is largely impossible. (see: [Why you can't catch a falling dollar](https://www.youtube.com/watch?v=AEpQ8YxupfQ)).

What I proposed, and is more than likely how every other VR dev is doing it, is to allow the user to _preload_ the grab, or anticipate it, by pulling the trigger when the object is within X units. If the falling object makes contact with the controller within that threshold (X), then they 'Catch' the object. If they pull the trigger too soon or of course too late and they miss.

For a practical example fire up *The Lab*, or pretty much any VR game/demo.

Cheers!

-Steven

**Comment by thestonefox** [comment]
I'll take a look at this maybe tomorrow.

I think I'll add it as an extra script you add to the controller that adds another trigger collider that's larger, then if the object falls into that collider then it will be set as touching, even though you're not touching it.

Maybe need some extra work on it where it only activates if you've thrown an item. Otherwise it may be weird for items that are stationary, so perhaps if the object you're colliding with has a velocity then it's catchable

**Comment by xEvoGx** [comment]
Well this really needs no work beyond simply allowing a _window_ of time where when the trigger is pulled and held, when an object comes in contact with the controller it will be *Grabbed*. In other words the moment the trigger/grip/whatever is pulled:

`canGrabFallingObject -= Time.deltaTime; `

```
if (canGrabFallingObject < 0)
{
     //Nothing happens!
}
else if (canGrabFallingObject > 0  &&  collidersTouch)
{
     //Grab that shit!
}
```

A larger collider would allow the falling object to be grabbed when it wasn't near the controller.

Finally **Rigidbody.Velocity > 0.0**?

**Comment by thestonefox** [comment]
Yeah I can give that way a go and see if it produces a good output.

**Comment by thestonefox** [comment]
Thinking about it, this way may not work.

It relies on the controller still touching the object at some point in the near future.

The problem may be being caused by collision tunnelling (where the object is moving so fast that the physics engine is moving it at too large a step for the collision to ever take place)

One solution to that is to enable continuous collision detection but it's a bit of a processor hog if it's on for lots of objects.

**Comment by xEvoGx** [comment]
Well _Continuous_ on the moving object(s) only?

Also to be clear, I can't see many instances where the thrown object is moving _that_ quickly. If you were trying to catch a bullet fired at you sure, but tossing a ball up?

**Comment by xEvoGx** [comment]
Say this becomes an issue as you say. We could combine your idea for a larger Collider around the controller when the trigger is pulled and < Time.deltaTime, and using **OverlapSphere** when the object is overlapping, turn on _Continuous_ and check for contact with the controller grab collider.

**Comment by thestonefox** [comment]
The collision tunneling is a slight issue in that the object's velocity is moving so fast that the rigidbody connects and is grabbable but the velocity takes it into an ungrabbed state so it gets dropped

I think i've solved it by just setting the object's velocity to zero on grab.

Not sure if this is a good idea to do all the time anyway (would you ever want to hold an object that keeps it's velocity? seems weird)

**Comment by xEvoGx** [comment]
Well right...it's being caught so Velocity should be zero no? Don't get me wrong...I'm not looking for authenticity in the physics sim, just in resultant interaction.

Well either way I'll be ready to test! :)

**Comment by thestonefox** [comment]
Yeah I agree, if you catch it, then it should lose all velocity. 

Like in the real world if you caught a fast moving heavy object, then you'd be pulled with that object till it's momentum was gone.

But in a VR game if you catch (let's say a tron disc) and it pulls your controller or play rig across the floor, then that's really jarring and doesn't work well.

I'll zero the velocity (and angularVelocity) whenever you grab anything (not just in the cooldown period catch)

**Comment by thestonefox** [comment]
Ok latest version has this in now. set the Grab Precognition variable on the Interact Grab script to set your window of grab time

**Comment by xEvoGx** [comment]
Perfect! As per usual I'll peep it when I'm back in front of the VR rig!

**Comment by xEvoGx** [comment]
Ugh I'm sorry I completely blanked on this!

Works fantastically, though not until I entered values around 100, which seems largely arbitrary from an end-user perspective (vs. using _seconds_ lets say). 

One odd thing that happened, is at times the object would stay attached to the controller without my having to hold the grip button. To release it, I had to grip again and then release. I can't say if this occurred only after playing catch with objects, but it happened nonetheless.

Also for clarity, I was tossing around the two cubes (green/red) in **021 - GrabbingObjectsWithJoints**, and any of the subsequent files that included that 'scene' with the drawer, box with lid and door.

Thanks Stone!

-Steven
